src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.cs

```diff
@@ -0,0 +1,477 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Globalization;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Completion;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Text.Shared.Extensions;
+using Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion;
+using Microsoft.VisualStudio.Text;
+using Roslyn.Utilities;
+using AsyncCompletionData = Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.Data;
+using FilterResult = Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.Controller.Session.FilterResult;
+using RoslynCompletionItem = Microsoft.CodeAnalysis.Completion.CompletionItem;
+using Session = Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.Controller.Session;
+using VSCompletionItem = Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.Data.CompletionItem;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.AsyncCompletion
+{
+    internal class ItemManager : ForegroundThreadAffinitizedObject, IAsyncCompletionItemManager
+    {
+        private readonly CompletionHelper _completionHelper;
+
+        private const int MaxMRUSize = 10;
+        private ImmutableArray<string> _recentItems = ImmutableArray<string>.Empty;
+
+        internal ItemManager(IThreadingContext threadingContext) : base(threadingContext)
+        {
+            _completionHelper = new CompletionHelper(isCaseSensitive: true);
+        }
+
+        public Task<ImmutableArray<VSCompletionItem>> SortCompletionListAsync(
+            IAsyncCompletionSession session,
+            AsyncCompletionData.AsyncCompletionSessionInitialDataSnapshot data,
+            CancellationToken cancellationToken)
+        {
+            AssertIsBackground();
+
+            SubscribeEvents(session);
```

Comment:
what is going on here?

Reply:
1. We'd like to update the list of most recent items on SessionCommitted. Here and in the update method we're subscribing for this event.
2. We'd like to unsubscribe from this event when session is dismissed. For the purpose, we also subscribe for SessionDismissed.


---
In reply to: [232509421](https://github.com/dotnet/roslyn/pull/29016#discussion_r232509421) [](ancestors = 232509421)